### PR TITLE
Add support for passing server params to factory

### PR DESCRIPTION
**Pull request by shadowhand**
The PSR-7 specification provides no way to update the server parameters
after a server request is instantiated, allow passing these parameters
into the factory method itself.

Having a server factory that creates from globals is pointless as an
interface, since this functionality is universal and can be applied to
any implementation that uses the server request factory.

Refs #18
Refs #20
Refs #21

### CODE REVIEW

**Comment by shadowhand** [comment]
Thoughts @http-interop/http-middleware-contributors ?

**Comment by barryvdh** [comment]
Why not keep the `createServerRequestFromGlobals` interface? Or do you want to leave that to the library/developers to implement? Looking at Guzzle, this doesn't seem so trivial as it seems:
https://github.com/guzzle/psr7/blob/64862e854f876bc5aee2996cab2f552db8586065/src/ServerRequest.php#L156-L181

``` php
public static function fromGlobals()
{
    $method = isset($_SERVER['REQUEST_METHOD']) ? $_SERVER['REQUEST_METHOD'] : 'GET';
    $headers = function_exists('getallheaders') ? getallheaders() : [];
    $uri = self::getUriFromGlobals();
    $body = new LazyOpenStream('php://input', 'r+');
    $protocol = isset($_SERVER['SERVER_PROTOCOL']) ? str_replace('HTTP/', '', $_SERVER['SERVER_PROTOCOL']) : '1.1';
    $serverRequest = new ServerRequest($method, $uri, $headers, $body, $protocol, $_SERVER);
    return $serverRequest
        ->withCookieParams($_COOKIE)
        ->withQueryParams($_GET)
        ->withParsedBody($_POST)
        ->withUploadedFiles(self::normalizeFiles($_FILES));
}
```

**Comment by Nyholm** [comment]
Im okey with dropping the createFromGlobals. But I do not understand the signature of `createServerRequest`. How would that be used? 

``` php
// index.php

include 'vendor/autoload.php';

$factory = new ServerRequestFactory();
$request = $factory-> createServerRequest(
  $_SERVER['REQUEST_METHOD'], 
  $_SERVER['SERVER_PORT'] == '80'?'http://':'https://' . $_SERVER['SERVER_NAME']. $_SERVER['REUEST_URI'],
  $_SERVER
);
```

I suggest only use a single array argument.

**Comment by shadowhand** [comment]
@barryvdh your example shows exactly why we don't need an interface for creating from globals, namely that there is nothing specific to an implementation. A single universal class can be written using the other PSR-17 factories. Basically, this:

``` php
class GlobalServerRequestFactory
{
    public function __construct(
        ServerRequestFactoryInterface $requestFactory,
        StreamFactoryInterface $streamFactory,
        UploadedFileFactoryInterface $uploadFactory
    ) {
        $this->requestFactory = $requestFactory;
        $this->streamFactory = $streamFactory;
        $this->uploadFactory = $uploadFactory;
    }

    public function fromGlobals()
    {
        $method = isset($_SERVER['REQUEST_METHOD']) ? $_SERVER['REQUEST_METHOD'] : 'GET';
        $uri = ($_SERVER['SERVER_PORT'] == '80' ? 'http://' : 'https://') . $_SERVER['SERVER_NAME'] . $_SERVER['REUEST_URI'];
        $headers = function_exists('getallheaders') ? getallheaders() : [];
        $protocol = isset($_SERVER['SERVER_PROTOCOL']) ? str_replace('HTTP/', '', $_SERVER['SERVER_PROTOCOL']) : '1.1';
        $body = $this->streamFactory->createFromResource(fopen('php://input', 'r+'));

        return $this->requestFactory->createServerRequest($method, $uri, $_SERVER)
            ->withBody($body)
            ->withCookieParams($_COOKIE)
            ->withQueryParams($_GET)
            ->withParsedBody($_POST)
            ->withUploadedFiles($this->normalizeFiles($_FILES));
    }

    private function normalizeFiles(array $files)
    {
        // also generic code, see guzzle implementation
    }
}
```

**Comment by shadowhand** [comment]
@Nyholm it certainly could be used that way. The main point about the `$server` parameter is that it is the only way to define the values for `ServerRequestInterface::getServerParams`, as defined in PSR-7:

``` php
    /**
     * Retrieve server parameters.
     *
     * Retrieves data related to the incoming request environment,
     * typically derived from PHP's $_SERVER superglobal. The data IS NOT
     * REQUIRED to originate from $_SERVER.
     *
     * @return array
     */
    public function getServerParams();
```

**Comment by shadowhand** [comment]
@Nyholm after thinking about your comment (over a cup of tea 🍵 !) I think I see your point. I can't really think of a scenario where the `$server` would not contain all of the necessary things to create the request. As such, maybe this is the minimum viable requirement:

``` php
$request = $serverRequestFactory->createServerRequest($_SERVER);
```

This can be mocked and fits well with the other factories, where only the minimum required parameters to create a valid object are defined.

Any other thoughts @barryvdh @mindplay-dk ?

**Comment by barryvdh** [comment]
If you're going for minimal required, then yes, you could leave out the rest. But from a user-experience point of view, you're not making it easier. Or do you plan to also provide that method as generic code next to the contracts?

> This can be mocked and fits well with the other factories, where only the minimum required parameters to create a valid object are defined.
> The signature for the normal request is `public function createRequest($method, $uri);`, which are both not required, because you can call `->withMethod()` and `withUri()` on the request, right?

From a UX view as a developer, most of the time I just want the ServerRequest from the globals (usually without changing). And while testing, I'm not sure what the common case is, but I would expect that you would often only need to define the method/uri. And then being able to define the method/uri directly is way more readable imho.
But from a pure technical perspective, it would be enough to just provider the $_SERVER array I guess..

**Comment by shadowhand** [comment]
@barryvdh `createRequest($method, $uri)` is the minimum viable signature to get a request object _that has valid state_. A request without these values is incomplete according to HTTP RFC.

Perhaps flipping the position of the `$server` parameter would be sufficient?

``` php
createServerRequest($server, $method = null, $uri = null);
```

Along with the assumption that if the method or URI is provided, it will be used instead of the value contained in `$server`?

**Comment by shadowhand** [comment]
@barryvdh and to answer your other question:

> Or do you plan to also provide that method as generic code next to the contracts?

Once the factory interfaces become stable, I see no reason not to provide a separate `http-global-server-factory` package that contains an implementation for creating the server request from globals. There are only small differences between Guzzle/Diactoros/Slim when it comes to this.

**Comment by weierophinney** [comment]
@shadowhand I think the last signature you provided is spot on, with one minor change:

``` php
createServerRequest(array $server, $method = null, $uri = null)
```

(`$server` MUST be an array.)

As you note, _generally_ you can derive the method and URI from `$_SERVER`, which means that the `$server` argument is the only _required_ argument in the majority of use cases, as it's the only value that cannot be derived and injected via other methods. Since the use case is to provide a bare instance which you then incrementally build via `with*()` methods, this makes a ton of sense.

**Comment by barryvdh** [comment]
So in that example, would the method/uri be derived from the $server param, unless the $method/$uri are filled? And would it throw an exception when that's not possible? eg. `$factory->createServerRequest([]);` would result in an invalid state.

**Comment by weierophinney** [comment]
> And would it throw an exception when that's not possible?

I think that makes the most sense. @shadowhand, thoughts?

**Comment by shadowhand** [comment]
@barryvdh Yes, the description should be updated to state that an exception should be thrown if the method and URI cannot be determined.

**Comment by shadowhand** [comment]
I have updated the PR as per the `fn(array server, method = null, uri = null)` signature and added the `@throws`.

File: src/ServerRequestFactoryInterface.php [hunk 1]
```diff
@@ -10,10 +10,20 @@
     /**
      * Create a new server request.
      *
-     * @param string $method
-     * @param UriInterface|string $uri
+     * The method and URI of the request SHOULD be derived from the given
+     * server parameters.
+     *
+     * If the method or URI parameter is passed, these values MUST be used
+     * instead of the server parameters.
```
Comment (shadowhand) [inline src/ServerRequestFactoryInterface.php:17]: @weierophinney I feel like this docblock could be worded better. Any suggestions?

Comment (weierophinney) [reply src/ServerRequestFactoryInterface.php:17]: I think it's succinct, and understandable immediately. I have no problems leaving it as you've worded it here.

**Review by weierophinney — APPROVED** [review]

**Review by Nyholm — APPROVED** [review]

**Comment by shadowhand** [comment]
This has been tagged in 0.2.0.

**Comment by mindplay-dk** [comment]
``` php
createServerRequest(array $server, $method = null, $uri = null)
```

How can URI and method be optional? You can't have a valid request object without method and URI, according to PSR-7 - `getMethod()` for example is annotated with `@return string` which would then be violated.

A request by definition is something that has a method and URI - if you don't have those values, you don't have a request.

**Comment by weierophinney** [comment]
@mindplay-dk Method and uri can be optional as they _could_ be derived from the `$server` array values (this is how we derive it typically when using the standard php SAPI implementations, for instance). The factory method should raise an exception if unable to derive these values and none were provided.

**Comment by mindplay-dk** [comment]
@weierophinney well, this further inflames my hatred for an HTTP model with so many forms of denormalization, but, yeah, under the circumstances, you are right :-/

**Pull request merged**